63.CUDAInterop/main.cpp

```diff
@@ -0,0 +1,189 @@
+			reinterpret_cast<float*>(cpubuffers[j]->getPointer())[i] = rand();
+
+	{
+		auto createBuffer = [&](core::smart_refctd_ptr<asset::ICPUBuffer>const& cpuBuf) {
+			auto buf = util->createFilledDeviceLocalBufferOnDedMem(queues[CommonAPI::InitOutput::EQT_COMPUTE],
+				{ {.size = size, .usage = asset::IBuffer::EUF_STORAGE_BUFFER_BIT | asset::IBuffer::EUF_TRANSFER_DST_BIT},
+				{{.externalMemoryHandType = video::IDeviceMemoryBacked::EHT_OPAQUE_WIN32}} },
+				cpuBuf->getPointer());
+			assert(buf.get());
+			return buf;
+		};
+
+		auto buf0 = createBuffer(cpubuffers[0]);
+		auto buf1 = createBuffer(cpubuffers[1]);
+		auto buf2 = createBuffer(cpubuffers[2]);
+		
+		video::CCUDADevice::SSharedCUDAMemory mem0, mem1, mem2;
+
+		ASSERT_SUCCESS(cudaDevice->importGPUBuffer(buf0.get(), &mem0));
+		ASSERT_SUCCESS(cudaDevice->importGPUBuffer(buf1.get(), &mem1));
+		ASSERT_SUCCESS(cudaDevice->importGPUBuffer(buf2.get(), &mem2));
+
+		void* parameters[] = { &mem0.ptr, &mem1.ptr, &mem2.ptr, &numElements };
+		ASSERT_SUCCESS(cu.pcuLaunchKernel(kernel, gridDim[0], gridDim[1], gridDim[2], blockDim[0], blockDim[1], blockDim[2], 0, stream, parameters, nullptr));
+		ASSERT_SUCCESS(cu.pcuMemcpyDtoHAsync_v2(cpubuffers[2]->getPointer(), mem2.ptr, size, stream));
+		ASSERT_SUCCESS(cu.pcuCtxSynchronize());
+
+		float* A = reinterpret_cast<float*>(cpubuffers[0]->getPointer());
+		float* B = reinterpret_cast<float*>(cpubuffers[1]->getPointer());
+		float* C = reinterpret_cast<float*>(cpubuffers[2]->getPointer());
+
+		for (auto i = 0; i < numElements; i++)
+			assert(abs(C[i] - A[i] - B[i]) < 0.01f);
+	}
+}
+
+
+void cuda2vk(
+	core::smart_refctd_ptr<video::CCUDAHandler> cudaHandler,
+	core::smart_refctd_ptr<video::CCUDADevice> cudaDevice,
+	video::IUtilities* util,
+	video::ILogicalDevice* logicalDevice,
+	nbl::video::IGPUQueue** queues,
+	CUfunction kernel,
+	CUstream stream,
+	int = 0)
+{
+	auto& cu = cudaHandler->getCUDAFunctionTable();
+
+	core::smart_refctd_ptr<asset::ICPUBuffer> cpubuffers[3] = { core::make_smart_refctd_ptr<asset::ICPUBuffer>(size),
+																core::make_smart_refctd_ptr<asset::ICPUBuffer>(size),
+																core::make_smart_refctd_ptr<asset::ICPUBuffer>(size) };
+	for (auto j = 0; j < 2; j++)
+		for (auto i = 0; i < numElements; i++)
+			reinterpret_cast<float*>(cpubuffers[j]->getPointer())[i] = rand();
+
+	{
+		video::CCUDADevice::SSharedCUDAMemory mem0, mem1, mem2;
+		ASSERT_SUCCESS(cudaDevice->createExportableMemory(size, sizeof(float), &mem0));
+		ASSERT_SUCCESS(cudaDevice->createExportableMemory(size, sizeof(float), &mem1));
+		ASSERT_SUCCESS(cudaDevice->createExportableMemory(size, sizeof(float), &mem2));
+
+		void* parameters[] = { &mem0.ptr, &mem1.ptr, &mem2.ptr, &numElements };
+		ASSERT_SUCCESS(cu.pcuMemcpyHtoDAsync_v2(mem0.ptr, cpubuffers[0]->getPointer(), size, stream));
+		ASSERT_SUCCESS(cu.pcuMemcpyHtoDAsync_v2(mem1.ptr, cpubuffers[1]->getPointer(), size, stream));
+		ASSERT_SUCCESS(cu.pcuCtxSynchronize());
```

Comment:
why synchronize before launching a kernel?

Reply:
I thought it was needed in host to device copies. I guess we don't need to sync as long as cpu buffers stay in scope.